packages/contractkit/src/wrappers/MetaTransactionWallet.ts

```diff
@@ -0,0 +1,289 @@
+    return this.executeMetaTransaction(tx, signature)
+  }
+
+  private getMetaTransactionDigestParams = (
+    tx: TransactionInput<any>,
+    nonce: number
+  ): [string, string, string, number] => {
+    const rawTx = this.toRawTransaction(tx)
+    return [rawTx.destination, rawTx.value, rawTx.data, nonce]
+  }
+
+  getMetaTransactionDigest = proxyCall(
+    this.contract.methods.getMetaTransactionDigest,
+    this.getMetaTransactionDigestParams,
+    stringIdentity
+  )
+
+  private getMetaTransactionSignerParams = (
+    tx: TransactionInput<any>,
+    nonce: number,
+    signature: Signature
+  ): [string, string, string, number, number, string, string] => {
+    const rawTx = this.toRawTransaction(tx)
+    return [
+      rawTx.destination,
+      rawTx.value,
+      rawTx.data,
+      nonce,
+      signature.v,
+      signature.r,
+      signature.s,
+    ]
+  }
+  getMetaTransactionSigner = proxyCall(
+    this.contract.methods.getMetaTransactionSigner,
+    this.getMetaTransactionSignerParams,
+    stringIdentity
+  )
+
+  eip712DomainSeparator = proxyCall(this.contract.methods.eip712DomainSeparator)
+  isOwner = proxyCall(this.contract.methods.isOwner)
+  nonce = proxyCall(this.contract.methods.nonce, undefined, valueToInt)
+  private getSigner = proxyCall(this.contract.methods.signer, undefined, stringIdentity)
+
+  transferOwnership: (newOwner: Address) => CeloTransactionObject<void> = proxySend(
+    this.kit,
+    this.contract.methods.transferOwnership
+  )
+
+  setSigner: (newSigner: Address) => CeloTransactionObject<void> = proxySend(
+    this.kit,
+    this.contract.methods.setSigner
+  )
+
+  setEip712DomainSeparator: () => CeloTransactionObject<void> = proxySend(
+    this.kit,
+    this.contract.methods.setEip712DomainSeparator
+  )
+
+  /**
+   * Get an cache the chain ID -- assume it's static for a kit instance
+   * @returns chainId
+   */
+  _chainId?: number
+  private async chainId(): Promise<number> {
+    if (this._chainId === undefined) {
+      this._chainId = await this.kit.web3.eth.net.getId()
+    }
+    return this._chainId
+  }
+
+  /**
+   * Get an cache the signer - it should be static for a Wallet instance
+   * @returns signer
+   */
+  _signer?: Address
+  public async signer() {
+    if (this._signer === undefined) {
+      this._signer = await this.getSigner()
+    }
+    return this._signer
+  }
+
+  /**
+   * Turns any possible way to pass in a tranasction into the raw values
+   * that are actually required. This is used both internally to normalize
+   * ways in which transactions are passed in but also public in order
+   * for one instance of ContractKit to serialize a meta transaction to
+   * send over the wire and be consumed somewhere else.
+   * @param tx TransactionInput<any> union of all the ways we expect transactions
+   * @returns a RawTransactions that's serializable
+   */
+  public toRawTransaction(tx: TransactionInput<any>): RawTransaction {
+    if ('destination' in tx) {
```

Comment:
The string based check seems error prone to me. Can we do something like this?
```suggestion
    if ((tx as RawTransaction).destination) {
```

Reply:
Could you elaborate on why the string check feels error prone?

Theoretically it wouldn't make sense to do a cast at that point because we're relying on the compiler's understanding of the object structure. So `tx` starts as being TransactionInput<any> which is a combination of 3 possible object structures. By checking whether props exist on the object we are discriminating between the types.

So the compile would complain if none of the branches had a `destination` prop and on the flip side it actually uses that if condition to discriminate, so inside of the first if statement `tx` becomes a `RawTransaction` automatically. 

So the point is we're letting the compiler tell us if our assertion that `destination` being a field of the input object is enough for it to determine that tx must be a `RawTransaction`. If one of the other branches were to add a `destination` field the compiler would start complaining but if we cast to `RawTransaction` we won't get that benefit and we'd be depriving the compiler of the ability to discriminate between fields.

---

Checking for the presence of a key feels ok to me as long as we're ensuring other data boundaries correctly. Aka we are not passing in manually cast data of uncertain structure - but that's true of any situation in which we're expecting a certain object structure and we might be shooting ourselves in the foot.

Reply:
Sorry for the late reply. I was worried that if `destination` string had a typo then the compiler wouldn't catch it as it would with the casting approach. However, based on what you're saying here, it sounds like that's not the case:
> compile would complain if none of the branches had a destination prop